### ISSUE TITLE
ENH: Support Sharded Repodata (CEP 16)

### ISSUE THREAD

**Issue by adrinjalali** [mamba-org/mamba#3431]
In an IRL conversation with @wolfv and others he mentioned the infra for avoiding the download of the massive package metadata when running `micromamba` commands is there, but I can't find a way to use it. Is there a way already to do that with a `micromamba` command?

To be specific, this is what I want to avoid:

```sh
$ mmamba install jupyter
conda-forge/noarch                                  16.3MB @   7.9MB/s  2.1s
conda-forge/linux-64                                37.6MB @   9.8MB/s  3.9s

...
```

**Comment by wolfv** [comment]
Sadly, no. It's only implemented in `rattler` / `pixi` for now. There you can point to `https://fast.prefix.dev/conda-forge` to get a speedy version.

**Comment by adrinjalali** [comment]
Is this completed? Or it's a "wontfix" kinda thing?

**Comment by Hind-M** [comment]
So at the moment it's not implemented in `mamba`/`micromamba`.
It could be at some point, but not in the short/medium term unfortunately.
We can also mark it as a feature request.

**Comment by jjerphan** [comment]
For some context, https://github.com/conda/ceps/pull/75/ formalizes sharded repodata, an improved indexing solution for `conda` channels.

It has been accepted, but it needs to be implemented by forges (including conda-forge), and by packages managers (such as conda and mamba/micromamba).

**Comment by jjerphan** [comment]
https://github.com/conda/conda/issues/14060 is the main requirement.

**Comment by jjerphan** [comment]
I would prioritize it, since UX-wise this is what cause latency and fatigue for users.

**Comment by baszalmstra** [comment]
Good to mention that the conda-forge and bioconda mirrors (but any channel really) at https://prefix.dev/ fully support shared repodata. So I think an implementation is not blocked on server support.

**Comment by jjerphan** [comment]
Yes, mamba can implement the support for the sharded repodata now without waiting for conda-lock to implement it.

**Comment by jjerphan** [comment]
I am not sure about `libsolv` being usable with the sharded repodata since `libsolv` has to load all the metadata of packages upfront AFAIK.

**Comment by baszalmstra** [comment]
Im not entirely sure how you "feed" the data to libsolv but we also have this working in rattler with our libsolv backend.

We preprocess the input specs and recursively fetch the packages for which we encounter depedencies. This is still an eager process but most of the time fetches only a fraction of the total number of packages from the channel. It does include all the records needed for a solve.

As an example, given just the spec "python", we start by fetching all records for python. We then iterate over all dependencies of all python records and see which package names we encounter. For instance `libsqlite`. We fetch all records of libsqlite and do the same there. This will crawl the entire search space that libsolv would needs to solve the input spec. 

You can also implement some smarts tricks here to shrink the total search space. Rattler also dispatch requests for different packages in parallel (make sure to use http/2) and we can aggressively cache the individual shards. This is what makes sharded repodata so fast.

You could also create `.solv` files for the individual shards for possibly even more speed! (We didnt implement this)

Let me know if you want more details, I can point you to where this happens in the code.

**Comment by jjerphan** [comment]
As of now in mamba, entire channels' `repodata.json` are loaded in `libsolv`'s DataBase, then a problem is constructed and solved as far as I remember.

**Comment by baszalmstra** [comment]
I see. Yeah that's a bit of a shortcoming as that won't scale very well. It means reading all the data ever created even if that is not needed at all. I also assume that will use _much_ more memory than needed..

**Comment by pchemguy** [comment]
This is a REAL PAIN. I do not need these updates AT ALL, unless specifically requested. Add `--use-index-cache` flag.

**Comment by h-vetinari** [comment]
It would be good to name this issue after the "sharded repodata" CEP (e.g. "ENH: support sharded repodata"), which is the most consistently used name for this feature IMO.

For reference, conda 25.11 has experimental support for sharded repodata based on `conda-libmamba-solver` [v25.11](https://github.com/conda/conda-libmamba-solver/releases/tag/25.11.0), so I think mamba is the last of the major three installers to not support this.

**Comment by jjerphan** [comment]
We agree that this is a significant design issue of mamba (which followed conda's original design), we likely will work on this issue if we have the budget and the focus to — this would necessitate to adapt the architecture of mamba quite significantly.

@pchemguy: your harsh imperative comment is unhelpful, yet your contributions are welcome especially if you want to follow your "passion for exploring coding".

**Comment by jjerphan** [comment]
A small teaser. 😉

![Comparison of resolving python with flat and sharded repodata](https://github.com/user-attachments/assets/3ce14298-8abe-40f0-bd6b-beae1e923fa6)

**Comment by pchemguy** [comment]
Not the best way to demo. There is enough time to see the first slow part, but I guess that is not what you are trying to demo. Getting anything that follows/flashes is virtually impossible.

**Comment by wolfv** [comment]
Nice @jjerphan!

**Comment by jjerphan** [comment]
@pchemguy: do you want a refund? 👼

**Comment by pchemguy** [comment]
> [@pchemguy](https://github.com/pchemguy): do you want a refund? 👼

Sure, why not :D

**Comment by jjerphan** [comment]
The message was meant to be a short appetizer and follow-up with timers, not to be the "_Grand Final Demo_".

Note that the developers of `mamba` are not at the ones behind the original design of indices used by `conda` channels which had to be adopted by `mamba` (and which has been constraining its UX, and even annoying its maintainers 😉). We are rectifying its UX when we have time and resources to work on it out of all others projects we improve and maintain.

**Comment by jjerphan** [comment]
Resolved in recent PRs such as #4145 #4148 #4149 #4161 #4162 #4167 #4174 #4177, #4174, #4187, #4192, #4195, #4190, #4206, #4207, #4212, #4220 (and potentially others).

**Comment by jezdez** [comment]
Amazing, thank you @jjerphan!

**Comment by jjerphan** [comment]
You are welcome! 2.6.0.rc0 should be out soon.

**Comment by jjerphan** [comment]
`conda-forge/label/micromamba_prerelease::micromamba==2.6.0.rc0` is out!
